### ISSUE TITLE
Add animation-require-prefers-reduced-motion (name pending)

### ISSUE THREAD

**Issue by mattxwang** [stylelint/stylelint#6202]
### What is the problem you're trying to solve?

Branching off of our discussion around accessibility features in @stylelint/contributors and a discussion with the FE accessibility team at @chanzuckerberg, one pattern we want to enforce is the use of `prefers-reduced-motion` in blocks that also define animations.

One positive use-case, taken from @chanzuckerberg's [design system](https://github.com/chanzuckerberg/edu-design-system/blob/main/src/upcoming-components/LoadingIndicator/LoadingIndicator.module.css)

```css
.loading-indicator__icon {
  animation: rotateIcon 2s linear infinite;

  @media screen and (prefers-reduced-motion) {
    animation: none;
  }
}
```

In contrast, we would like to disallow 

```css
.loading-indicator__icon {
  animation: rotateIcon 2s linear infinite;
  /* error: prefers-reduced-motion is not used here */
}
```

A similar rule could apply to `transition-*` keywords as well.

Two caveats:

1. I am not sure if this is within scope of core Stylelint. If this is better suited for [stylelint-accessible](https://github.com/thibaudcolas/stylelint-accessible) or somewhere else (e.g. another custom plugin), please let me know!
2. I am not sure if this is technically feasible, not performant, or would trigger too many false positives/negatives.

### What solution would you like to see?

A rule to enforce the behaviour explained above. I'm no naming expert, so help there is certainly welcome!

With the caveat that I have little experience writing rule specs, I'm thinking something like:

- Name: `animation-require-prefers-reduced-motion` (name pending)
- Primary option: none (would we want to add some sort of `ignoreSelectors` feature here?
- Secondary options: none
- Autofixable: No
- Message: - `Expected `prefers-reduced-motion` to be used within declaration block `
- Description: "Enforce the use of `prefers-reduced-motion` when defining animations and transitions."
- Extended description: more polished text about the importance of `prefers-reduced-motion` and accessibility
- Section: not sure - would we create a new subsection under Enforce conventions?

I'm happy to work on an implementation pending the discussion on this rule!

**Comment by mattxwang** [comment]
cc: @thibaudcolas (is this what you meant by correct use of `prefers-reduced-motion` and @dierat for the rule suggestion within CZI.

**Comment by thibaudcolas** [comment]
@mattxwang yup. See also [stylelint-a11y’s rule](https://github.com/YozhikM/stylelint-a11y/tree/master/src/rules/media-prefers-reduced-motion).

Another common way to implement `prefers-reduced-motion` is to remove all animations everywhere at once:

```css
@media (prefers-reduced-motion: reduce) {
  *, ::before, ::after {
    animation-delay: -1ms !important;
    animation-duration: 1ms !important;
    animation-iteration-count: 1 !important;
    background-attachment: initial !important;
    scroll-behavior: auto !important;
    transition-duration: 0s !important;
    transition-delay: 0s !important;
  }
}
```

I don’t think this undermines the usefulness of the rule as you suggested it, but it’s worth keeping in mind – at least the appropriateness of the rule to different approaches would need to be documented.

We also need to be careful with our examples (as this is what people will refer to when trying to figure out what to make of an error):

- You have `@media screen and (prefers-reduced-motion)`, I would have expected `@media (prefers-reduced-motion: reduce)`. Not sure which makes the most sense / what the differences are.
- `animation: none` would break any JS code relying on animation events (hence why my example above makes animations "instantaneous" rather than removing them completely).

**Comment by ybiquitous** [comment]
Thanks for the exciting proposal. The idea of requiring the `prefers-reduced-motion` media feature sounds good to me, but it seems we clarify the purpose of the rule.

The ["About rules"](https://stylelint.io/user-guide/rules/about) doc says that built-in rules "have a singular purpose", but what properties should the new rule be applied to?

- `animation`
- `animation-*` (e.g. `animation-duration`)
- `transition`
- `transition-*` (e.g. `transition-duration`)
- ...and more?

Also, what name should be suitable for the new rule, according to the [naming guideline](https://stylelint.io/user-guide/rules/about#names)?

**Comment by mattxwang** [comment]
> See also [stylelint-a11y’s rule](https://github.com/YozhikM/stylelint-a11y/tree/master/src/rules/media-prefers-reduced-motion).

Ah, I think I should've done a bit more homework before making this issue. It does seem like it fulfills the requirements that I'm suggesting (or at least - mostly. I'm not sure if I like the autofix). This is maybe part of the broader discussion about bringing either that plugin and/or @thibaudcolas's work into the main repo/namespace. @thibaudcolas, did you want to write up that issue based on our conversation in the contributors team, or did you want me to?

> The ["About rules"](https://stylelint.io/user-guide/rules/about) doc says that built-in rules "have a singular purpose", but what properties should the new rule be applied to?

I personally think all of `animation`, `transition`, and `animation-*`/`transition-*` should all be within scope, similar to the stylelint-a11y rule.

> Also, what name should be suitable for the new rule, according to the [naming guideline](https://stylelint.io/user-guide/rules/about#names)?

Thanks for pointing that out. Looking at "No rules", I think our rule *just disallows* something, and so probably needs `*-no-*`. Is there a unified word that describes animations and transitions in the spec?

**Comment by nosilleg** [comment]
If the rule is looking for the existence of `prefers-reduced-motion` then `*-no-unreduced-motion` might be reasonable.

If a11y is going to be a part of core then maybe `a11y-no-unreduced-motion` works?
Closer to implementation instead of intent would be `media-no-unreduced-motion`.

I haven't looked at how difficult it is to limit the errors to "motion" based styles, but limiting it to vestibular based triggers I'm guessing is impossible.  As such there will always be false positives.

The stylelint-a11y autofix is definitely a "no animation" instead of "reduced motion" implementation.  I don't think this rule should be autofixable.
